### DISCUSSION TITLE
Avoid duplicating state registry entries but keep a simple interface?

### DISCUSSION THREAD

**Discussion by jaidetree** [pmndrs/jotai#1972] [Q&A] [answered]
Trying to model state and actions for a important, but small feature in our app. After setting up `useAtomsDevtools()` I realized it was registering 2 copies of the state in the registry.

First, there's the stateAtom, then the derivedWritableAtom booksStateAtom that uses actions to update the stateAtom, sometimes asynchronously, sometimes synchronously. 

```js
const stateAtom = atom({ books: [], fetching: false })
const booksStateAtom = atom(
  get => get(stateAtom),
  async (get, set, action) => {
    const state = get(stateAtom)
    switch (action.type) {
      case 'init': {
        set(stateAtom, { ...state, fetching: true })
        const books = await fetchBooks(action.id)
        
        set(stateAtom, { ...state, books, fetching: false })
        return
      },

      case 'add_book': {
        const { book } = action
       
        set(stateAtom, { ...sate, books: [...state.books, book] })
      }
    }
  })
```

Usage:

```tsx
function SomeComponent () {
  const [state, dispatch] = useAtom(booksStateAtom)
  
  return (
    <div>{/* ... */}</div>
  )
}
```

While that works for the most part, it turns out that it needs to store 2 copies of the stateAtom:

```js
{
  atom1: { fetching: false, books: [] },
  atom2: { fetching: false, books: [] },
}
```

Then I changed the general booksStateAtom to dispatchAtom:

```js
const dispatchAtom = atom(
  null,
  async (get, set, action) => {
    const state = get(stateAtom)
    switch (action.type) {
      case 'init': {
        set(stateAtom, { ...state, fetching: true })
        const books = await fetchBooks(action.id)
        
        set(stateAtom, { ...state, books, fetching: false })
        return
      },

      case 'add_book': {
        const { book } = action
       
        set(stateAtom, { ...sate, books: [...state.books, book] })
      }
    }
  })
```

This ensures there's only 1 copy of the state in the registry but it makes using it a bit awkward.

The following no longer works:
```js
const [state, dispatch] = useAtom(booksStateAtom)
```

Which needs to become:
```js
const state = useAtomValue(stateAtom)
const dispatch = useSetAtom(dispatchAtom)
```

While not the end of the world, feels clunky, and for our team members mostly coming with hook experience, this is a little more unfamiliar and requires they understand the difference between those hooks and why the dispatchAtom must be separate.

I worked around it with a custom hook:

```ts
function useBookState () {
  return [useAtomValue(stateAtom), useSetAtom(dispatchAtom] as const
}
```

But it feels like this extra thing to be aware of, and an extra problem I need to solve because of Jotai.

Is there a better solution that I'm missing?

**Comment by dai-shi** [comment]
1. `get => get(stateAtom)` is often used technique. While it make two entries, it shouldn't be a big issue. (it depends.)
2. `dispatchAtom` is a nice solution for it. So, you are right on track.
3. In your specific use case, `atomWithReducer` may help.

**Reply by jaidetree** [reply]
Thanks what are the dependent circumstances where two copies of state would be an issue?

In my case, the response from the api server is quite large. Would that be a consideration?

**Reply by dai-shi** [reply]
The size of atom value shouldn't be a problem, because it doesn't copy the value. The circumstance would be when you have way too many atoms. But, I suspect Jotai won't become the bottleneck. Anyway, hypothetically, it's better to avoid duplicating if it doesn't sacrifices UX/DX.

**Reply by jaidetree** [reply]
I see, that makes sense. Was not sure with the immutability if the other atom was storing a reference to the object, or a separate copy.

That said, I did look at the atomWithReducer implementation as I was curious how you solved that problem.

```ts
  const anAtom: any = atom(initialValue, (get, set, action: Action) =>
    set(anAtom, reducer(get(anAtom), action))
  )
  return anAtom
```

That was exactly what I was missing! That was exactly what I was looking for:

- atom(initialValue) - Allows the set function to write to itself like `set(anAtom, ...)`
- atom(get => get(...)) - Makes it a derivative atom

I updated my code to use a similar pattern, and it seems to be working great!

```js
const stateAtom = atom(
  { books: [], fetching: false },
  async (get, set, action) => {
    const state = get(stateAtom)
    switch (action.type) {
      case 'init': {
        set(stateAtom, { ...state, fetching: true })
        const books = await fetchBooks(action.id)
        
        set(stateAtom, { ...state, books, fetching: false })
        return
      },

      case 'add_book': {
        const { book } = action
       
        set(stateAtom, { ...sate, books: [...state.books, book] })
      }
    }
  })
```

**Reply by jaidetree** [reply]
Anyway, thanks a ton for your help! I'll donate since this was something I could have answered for myself if I inspected the codebase more thoroughly 😄

**Reply by dai-shi** [reply]
As public API, `atom(initialValue)`, `atom(read)`, `atom(read, write)` are supported and documented.
But, technically, `atom(initialValue, write)` works too. The reason it's not documented is it can't be typed in TS.
So, as a workaround, `atomWithReducer` is provided to cover typical usage with type safety.

**Reply by jaidetree** [reply]
Thanks for the heads up. Just checked the CI results and sure enough the typechecks all failed around anything that touches that atom and some of the internal update logic. 

Practically, sounds like I'll need to make a wrapper function that's similar and if I can't do that, just go back to the separate state atom with derived value, as you said it's just going to be a pointer. I do need the ability to change state before async work begins, and the ability to do it after so the reducer util is not a perfect fit.

I take it the type issue is that TypeScript will never understand that calling `set` on the atom itself behaves differently than when calling it against other atoms?

**Reply by dai-shi** [reply]
> I take it the type issue is that TypeScript will never understand that calling `set` on the atom itself behaves differently than when calling it against other atoms?

yeah.

**Reply by jaidetree** [reply, marked as answer]
Came up with a solution: Create a local `setSelf` function that calls `set(...)` on itself.

```ts
const stateAtom = atom<State, [Action], void>(
  { books: [], fetching: false },
  async (get, set, action) => {
    const state = get(stateAtom)
    const setSelf: (state: State) => void = state => set(stateAtom, state as any)

    switch (action.type) {
      case 'init': {
        setSelf({ ...state, fetching: true })
        const books = await fetchBooks(action.id)
        
        setSelf({ ...state, books, fetching: false })
        return
      },

      case 'add_book': {
        const { book } = action
       
        setSelf({ ...state, books: [...state.books, book] })
        return
      }
    }
  })
```
